### PR TITLE
3 c analysis

### PR DESCRIPTION
**Pull request by jwassermann**
Added several techniques to compute polarization properties of 3C seismic records...

### CODE REVIEW

**Comment by krischer** [comment]
Still lacks a tutorial and maybe a particle plot but otherwise ready to review once Travis is finished.

**Comment by megies** [comment]
Should we drop in the deprecated kwarg decorator here? https://github.com/obspy/obspy/pull/505/files#diff-8fb1353dd0f92d5322bb3e69786e4c9aR26

It might break code.. and the name change is not really necessary..

File: misc/docs/source/citations.rst [hunk 1]
```diff
@@ -8,6 +8,140 @@
...
+.. [Withers1998]  
+   | Withers, M., Aster, R., Young, C., Beiriger, J., Harris, M., Moore, S., and Trujillo, J. (1998),
+   | **A comparison of select trigger algorithms for automated global seismic phase and event detection**,
+   | *Bulletin of the Seismological Society of America*, 88 (1), 95-106.
+   | http://www.bssaonline.org/content/88/1/95.abstract
+.. [Young1996]  
+   | Young, J. B., Presgrave, B. W., Aichele, H., Wiens, D. A., and Flinn, E. A. (1996),
+   | **The Flinn-Engdahl regionalisation scheme: the 1995 revision**,
+   | *Physics of the Earth and Planetary Interiors*, 96 (4), 223-297.
+=======
```
Comment (QuLogic) [inline misc/docs/source/citations.rst]: Is this supposed to be here? It looks like an incorrect merge.

File: obspy/signal/polarization.py [hunk 2]
```diff
@@ -32,53 +32,54 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
-        dimension window number, second dimension the actualy data.
-    :type datay: :class:`~numpy.ndarray`
-    :param datay: Data of y component. See also note in datax.
-    :type dataz: :class:`~numpy.ndarray`
-    :param dataz: Data of z component. See also note in datax.
-    :type fk: list
-    :param fk: Coefficients of polynomial used for calculating the time
+        dimension is the window number, second dimension is the data.
+    :type data_x: :class:`~numpy.ndarray`
+    :param data_y: Data of y component. See description of data_x.
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Place `data_x` in double-backticks.

File: obspy/signal/polarization.py [hunk 3]
```diff
@@ -32,53 +32,54 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    :type data_y: :class:`~numpy.ndarray`
+    :param data_z: Data of z component. See description of data_x.
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Place `data_x` in double-backticks.

File: obspy/signal/polarization.py [hunk 4]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+        default is set to 0.
+    :type noise_thres: float
+    :returns:  azimuth, incidence, rectilinearity, and planarity
+    """
+    Z = []
+    N = []
+    E = []
+    comp, npts = np.shape(stream)
+
+    for i in range(0, npts):
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Can use some vector operations here.

File: obspy/signal/polarization.py [hunk 5]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+        if (stream[0][i] ** 2 + stream[1][i] ** 2 + stream[2][i] ** 2) \
+                > noise_thres:
+            Z.append(stream[0][i])
+            N.append(stream[1][i])
+            E.append(stream[2][i])
+
+    X = np.vstack((E, N, Z))
+    covmat = np.cov(X)
+    eigvec, eigenval, v = (np.linalg.svd(covmat))
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Unnecessary parentheses.

File: obspy/signal/polarization.py [hunk 6]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    """
+    Simple function to estimate the instantaneous frequency based on the
+    derivative of the data and the analytical (hilbert) data.
+
+    :param data: The data array.
+    :type data: :class:`numpy.ndarray`
+    :param sampling_rate: The sampling rate in Hz.
+    :type sampling_rate: float
+    """
+    x = data.copy()
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: I'm not sure why this copy is necessary.

File: obspy/signal/polarization.py [hunk 7]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+
+
+def vidale_adapt(stream, noise_thres, fs, flow, fhigh, spoint, stime, etime):
+    """
+    Adaptive window polarization analysis after [Vidale1986]_ with the
+    modification of adapted analysis window estimated by estimating the
+    instantaneous frequency. It returns the azimuth, incidence, rectilinearity
+    planarity and ellipticity.
+
+    :param stream: ZNE sorted trace data
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Why does it need to be sorted? Isn't that what the `stream.sort` call does below?

File: obspy/signal/polarization.py [hunk 8]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    :type stream: :class:`~obspy.core.stream.Stream`
+    :param noise_thres: Variance of the noise sphere; data points are excluded
+        when falling within the sphere with radius sqrt(noise_thres),
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: s/with /of /

File: obspy/signal/polarization.py [hunk 9]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+
+def flinn(stream, noise_thres=0):
+    """
+    Computes the azimuth, incidence, rectilinearity and planarity after the
+    eigenstructure decomposition method of [Flinn1965b]_.
+
+    :param stream: ZNE sorted trace data
+    :type stream: :class:`~obspy.core.stream.Stream`
+    :param noise_tresh: Variance of the noise sphere; data points are excluded
+        when falling within the sphere with radius sqrt(noise_thres),
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: s/with /of /

File: obspy/signal/polarization.py [hunk 10]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+
+    :param stream: ZNE sorted trace data
+    :type stream: :class:`~obspy.core.stream.Stream`
+    :param noise_thres: Variance of the noise sphere; data points are excluded
+        when falling within the sphere with radius sqrt(noise_thres),
+        Default = 0
+    :type noise_thres: float
+    :param fs: sampling rate
+    :type fs: float
+    :param flow: lower frequency for analysis
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Add "limit" like next parameter.

File: obspy/signal/polarization.py [hunk 11]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    :type flow: float
+    :param fhigh: upper frequency limit for analysis
+    :type fhigh: float
+    :param spoint: array with traces` individual start times in sample
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Backtick should be an apostrophe, I think.

Comment (QuLogic) [inline obspy/signal/polarization.py]: s/sample/samples/

File: obspy/signal/polarization.py [hunk 12]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    :type stime: :class:`~obspy.core.utcdatetime.UTCDateTime`
+    :param etime: end time for the analysis
+    :type etime: :class:`~obspy.core.utcdatetime.UTCDateTime`
+    :returns: azimuth, incidence, rectilinearity planarity, and ellipticity
+    """
+    W = 3.0
+    stream.sort(reverse=True)
+    Z = stream[0].data.copy()
+    N = stream[1].data.copy()
+    E = stream[2].data.copy()
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: I'm not entirely sure why these `copy`s are necessary.

File: obspy/signal/polarization.py [hunk 13]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+
+    Zi = instantaneous_frequency(Z, fs)
+    Za = signal.hilbert(Z)
+    Ni = instantaneous_frequency(N, fs)
+    Na = signal.hilbert(N)
+    Ei = instantaneous_frequency(E, fs)
+    Ea = signal.hilbert(E)
+    res = []
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: You can probably pre-allocate this array.

File: obspy/signal/polarization.py [hunk 14]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    :type flow: float
+    :param fhigh: upper frequency limit for analysis
+    :type fhigh: float
+    :param spoint: array with traces` individual start times in sample
+    :type spoint: :class:`numpy.ndarray`
+    :param stime: start time of the analysis
+    :type stime: :class:`~obspy.core.utcdatetime.UTCDateTime`
+    :param etime: end time for the analysis
+    :type etime: :class:`~obspy.core.utcdatetime.UTCDateTime`
+    :returns: azimuth, incidence, rectilinearity planarity, and ellipticity
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: As a list of `np.ndarray`, which is a bit odd.

File: obspy/signal/polarization.py [hunk 15]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    offset = int(3 * fs / flow)
+    while True:
+        adapt = int(3. * W * fs / (Zi[offset] + Ni[offset] + Ei[offset]))
+        # in order to account for errors in the inst freq estimation
+        if adapt > int(3 * fs / flow):
+            adapt = int(3 * fs / flow)
+        if adapt < int(3 * fs / fhigh):
+            adapt = int(3 * fs / fhigh)
+        adapt /= 2
+        adapt = (2 * adapt) + 1
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Huh, divide by 2 and multiply by 2. Is this expecting integer division?

File: obspy/signal/polarization.py [hunk 16]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+        covmat[0][1] = np.dot(Ex, Nx.conjugate())
+        covmat[1][0] = covmat[0][1].conjugate()
+        covmat[0][2] = np.dot(Ex, Zx.conjugate())
+        covmat[2][0] = covmat[0][2].conjugate()
+        covmat[1][1] = np.dot(Nx, Nx.conjugate())
+        covmat[1][2] = np.dot(Zx, Nx.conjugate())
+        covmat[2][1] = covmat[1][2].conjugate()
+        covmat[2][2] = np.dot(Zx, Zx.conjugate())
+
+        eigvec, eigenval, v = (np.linalg.svd(covmat))
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Unnecessary parentheses.

Comment (QuLogic) [inline obspy/signal/polarization.py]: Except for the `ellip` calculation, from here down appears to be quite similar to `flinn`. Might be useful to put that in a common function.

File: obspy/signal/polarization.py [hunk 17]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+                int(spoint[2] + offset + adapt / 2)]
+        Nx = Na[int(spoint[1] + offset - adapt / 2):
+                int(spoint[1] + offset + adapt / 2)]
+        Ex = Ea[int(spoint[0] + offset - adapt / 2):
+                int(spoint[0] + offset + adapt / 2)]
+        Zx -= Zx.mean()
+        Nx -= Nx.mean()
+        Ex -= Ex.mean()
+
+        covmat = np.zeros([3, 3], dtype=np.complex128)
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: I would do this allocation outside the loop (though it's not too large.)

File: obspy/signal/polarization.py [hunk 18]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+
+        final = fminbound(fun, 0.0, math.pi, full_output=True)
+        X = 1. - final[1]
+        ellip = math.sqrt(1.0 - X ** 2) / X
+        # rectilinearity defined after Montalbetti & Kanasewich, 1970
+        rect = 1. - np.sqrt(eigenval[1] / eigenval[0])
+        # planarity defined after Jurkevics, 1988
+        plan = 1. - (2.0 * eigenval[2] / (eigenval[1] + eigenval[0]))
+
+        azimuth = 180 * math.atan2(eigvec[0][0].real, eigvec[1][0].real) / \
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: The only difference with `flinn` is this one doesn't use the `degrees` function.

File: obspy/signal/polarization.py [hunk 19]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+        when falling within the sphere with radius sqrt(noise_thres)
+    :type noise_thres: float
+    :returns: azimuth, incidence, error of azimuth, error of incidence
+    """
+    Z = []
+    N = []
+    E = []
+    comp, npts = np.shape(stream)
+
+    for i in range(0, npts):
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Probably can use some vector operations here.

File: obspy/signal/polarization.py [hunk 20]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+        if (stream[0][i] ** 2 + stream[1][i] ** 2 + stream[2][i] ** 2) \
+                > noise_thres:
+            Z.append(stream[0][i])
+            N.append(stream[1][i])
+            E.append(stream[2][i])
+
+    def fit_func(beta, x):
+        # Eventually this is correct: return beta[0] * x + beta[1]
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: What does this mean?

File: obspy/signal/polarization.py [hunk 21]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    out = odr.run()
+    az_slope = out.beta[0]
+    az_error = out.sd_beta[0]
+
+    N = np.asarray(N)
+    E = np.asarray(E)
+    Z = np.asarray(Z)
+    R = np.sqrt(N ** 2 + E ** 2)
+
+    data = scipy.odr.Data(R, np.sqrt(Z ** 2))
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Huh? `np.sqrt(Z ** 2)`?

File: obspy/signal/polarization.py [hunk 22]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    model = scipy.odr.Model(fit_func)
+    odr = scipy.odr.ODR(data, model, beta0=[1.0])
+    out = odr.run()
+    in_slope = out.beta[0]
+    in_error = out.sd_beta[0]
+
+    azim = math.atan2(1.0, az_slope)
+    inc = math.atan2(1.0, in_slope)
+    az_error = 1.0 / ((1.0 ** 2 + az_slope ** 2) * azim) * az_error
+    in_error = 1.0 / ((1.0 ** 2 + in_slope ** 2) * inc) * in_error
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Can drop the `**2` on the `1.0`.

File: obspy/signal/polarization.py [hunk 23]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+
+    return math.degrees(azim), math.degrees(inc), az_error, in_error
+
+
+def get_s_point(stream, stime, etime):
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Internal?

File: obspy/signal/polarization.py [hunk 24]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    slatest = stream[0].stats.starttime
+    eearliest = stream[0].stats.endtime
+    for tr in stream:
+        if tr.stats.starttime >= slatest:
+            slatest = tr.stats.starttime
+        if tr.stats.endtime <= eearliest:
+            eearliest = tr.stats.endtime
+
+    nostat = len(stream)
+    spoint = np.empty(nostat, dtype=np.int32, order="C")
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: `order` is pretty unnecessary for a 1D array.

File: obspy/signal/polarization.py [hunk 25]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    epoint = np.empty(nostat, dtype=np.int32, order="C")
+    # now we have to adjust to the beginning of real start time
+    if slatest > stime:
+        msg = "Specified start-time is smaller than starttime in stream"
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: s/start-time/start time/
s/smaller/earlier/
s/starttime/start time/  (maybe add in latest as well)

File: obspy/signal/polarization.py [hunk 26]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+        raise ValueError(msg)
+    if eearliest < etime:
+        msg = "Specified end-time bigger is than endtime in stream"
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: s/end-time/end time/
s/bigger/later/
s/endtime/end time/ (and maybe add in earliest as well)

File: obspy/signal/polarization.py [hunk 27]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+        raise ValueError(msg)
+    for i in range(nostat):
+        offset = int(((stime - slatest) / stream[i].stats.delta + 1.))
+        negoffset = int(((eearliest - etime) / stream[i].stats.delta + 1.))
+        diffstart = slatest - stream[i].stats.starttime
+        frac, ddummy = math.modf(diffstart)
+        spoint[i] = int(ddummy)
+        if frac > stream[i].stats.delta * 0.25:
+            msg = "Difference in start times exceeds 25% of samp rate"
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: s/samp/sampling/

File: obspy/signal/polarization.py [hunk 28]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    spoint = np.empty(nostat, dtype=np.int32, order="C")
+    epoint = np.empty(nostat, dtype=np.int32, order="C")
+    # now we have to adjust to the beginning of real start time
+    if slatest > stime:
+        msg = "Specified start-time is smaller than starttime in stream"
+        raise ValueError(msg)
+    if eearliest < etime:
+        msg = "Specified end-time bigger is than endtime in stream"
+        raise ValueError(msg)
+    for i in range(nostat):
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: `for i, tr in enumerate(stream):` -- All these `stream[i]` are too long...

File: obspy/signal/polarization.py [hunk 29]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+        diffstart = slatest - stream[i].stats.starttime
+        frac, ddummy = math.modf(diffstart)
+        spoint[i] = int(ddummy)
+        if frac > stream[i].stats.delta * 0.25:
+            msg = "Difference in start times exceeds 25% of samp rate"
+            warnings.warn(msg)
+        spoint[i] += offset
+        diffend = stream[i].stats.endtime - eearliest
+        frac, ddummy = math.modf(diffend)
+        epoint[i] = int(ddummy)
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Aren't these two lines just basically `int(diffend)`?

File: obspy/signal/polarization.py [hunk 30]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+        epoint[i] += negoffset
+
+    return spoint, epoint
+
+
+def polarization_analysis(stream, win_len, win_frac, frqlow, frqhigh, stime,
+                          etime, verbose=False, timestamp="mlabday",
+                          method="pm", var_noise=0.0):
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: I'm not sure why all these names need to be so shortened.

File: obspy/signal/polarization.py [hunk 31]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+        msg = "in array sampling rates of traces in stream are not equal"
+        raise ValueError(msg)
+
+    if verbose:
+        print("stream contains following traces:")
+        print(stream)
+        print("stime = " + str(stime) + ", etime = " + str(etime))
+
+    # offset of arrays
+    spoint, _epoint = get_s_point(stream, stime, etime)
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Is this not the only caller of `get_s_point`? Why bother calculating `epoint` then?

File: obspy/signal/polarization.py [hunk 32]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    # loop with a sliding window over the dat trace array and apply bbfk
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: It's neither `dat` nor `bbfk`, is it?

File: obspy/signal/polarization.py [hunk 33]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    fs = stream[0].stats.sampling_rate
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Already assigned at #L463.

File: obspy/signal/polarization.py [hunk 34]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    if method.lower() == "vidale":
+        res = vidale_adapt(stream, var_noise, fs, frqlow, frqhigh, spoint,
+                           stime, etime)
+    else:
+        nsamp = int(win_len * fs)
+        nstep = int(nsamp * win_frac)
+        newstart = stime
+        # 0.22 matches 0.2 of historical C bbfk.c
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Is there a reason to match historical C `bbfk.c`? Isn't this new code entirely?

File: obspy/signal/polarization.py [hunk 35]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+                print(newstart, (newstart + (nsamp / fs)), res[-1][1:])
+            offset += nstep
+
+            newstart += nstep / fs
+    res = np.array(res)
+    if timestamp == "julsec":
+        pass
+    elif timestamp == "mlabday":
+        # 719162 == hours between 1970 and 0001
+        res[:, 0] = res[:, 0] / (24. * 3600) + 719162
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Comment and constant are incorrect. See #956.

File: obspy/signal/polarization.py [hunk 36]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+    else:
+        msg = "Option timestamp must be one of 'julsec', or 'mlabday'"
+        raise ValueError(msg)
+
+    npt = len(res[:, 0])
+    npt /= 2
+
+    if method == "pm":
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Everywhere else checks `method.lower()`.

File: obspy/signal/polarization.py (hunk 35 above)
Comment (QuLogic) [inline obspy/signal/polarization.py]: Actually, why even bother? `res[:, 0]` is not returned.

File: obspy/signal/polarization.py [hunk 37]
```diff
@@ -86,48 +87,482 @@ def eigval(datax, datay, dataz, fk, normf=1):
...
+                    res.append(np.array([newstart.timestamp + nsamp / fs,
+                                         azimuth, incidence, error_az,
+                                         error_inc]))
+            if method.lower() == "flinn":
+                azimuth, incidence, reclin, plan = flinn(data, var_noise)
+                res.append(np.array([newstart.timestamp + nsamp / fs, azimuth,
+                                     incidence, reclin, plan]))
+
+            if verbose:
+                print(newstart, (newstart + (nsamp / fs)), res[-1][1:])
```
Comment (QuLogic) [inline obspy/signal/polarization.py]: Unnecessary parentheses.

File: obspy/signal/polarization.py (hunk 4 above)
Comment (mbyt) [inline obspy/signal/polarization.py]: Not easily, length is flexible. This means we would need two passes of the for loop

Comment (QuLogic) [inline obspy/signal/polarization.py]: Yea, but they'd be in C. Not sure if it'd be faster, though...
```
mask = (stream[0].data**2 + stream[1].data**2 + stream[2].data**2) > noise_thres
Z = stream[0].data[mask]
N = stream[1].data[mask]
E = stream[2].data[mask]
```

File: obspy/signal/polarization.py (hunk 22 above)
Comment (mbyt) [inline obspy/signal/polarization.py]: but possibly easier to understand

**Comment by mbyt** [comment]
Either we merge now, or never. Is a pull request from January 2013, that is 2 years old.

I'd vote for now.

**Pull request merged**